### PR TITLE
[Weighted Routing] Add support to allow custom string search

### PR DESCRIPTION
**Pull request by anshu1106**
### Description
This PR adds support to allow searches with custom string when strict weighed routing is enabled

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/6297

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java [hunk 1]
```diff
@@ -629,6 +654,13 @@ public ShardIterator preferNodeActiveInitializingShardsIt(Set<String> nodeIds) {
         return new PlainShardIterator(shardId, preferred);
     }
 
+    public boolean isNodeWeighedAway(WeightedRouting weightedRouting, DiscoveryNodes nodes, String localNodeId, double defaultWeight) {
+        String attVal = nodes.get(localNodeId).getAttributes().get(weightedRouting.attributeName());
+        // If weight for a zone is not defined, considering it as 1 by default
+        Double weight = weightedRouting.weights().getOrDefault(attVal, defaultWeight);
+        return weight.intValue() == WeightedRoutingMetadata.WEIGHED_AWAY_WEIGHT;
+    }
+
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java:663]: we are not using this . So are including nodes with weight 0 as well ?

File: server/src/main/java/org/opensearch/cluster/routing/OperationRouting.java [hunk 2]
```diff
@@ -343,7 +344,14 @@ private ShardIterator preferenceActiveShardIterator(
...
         // shard ID into the hash of the user-supplied preference key.
         routingHash = 31 * routingHash + indexShard.shardId.hashCode();
-        if (ignoreAwarenessAttributes()) {
+        if (weightedRoutingMetadata != null && weightedRoutingMetadata.getWeightedRouting().isSet() && isStrictWeightedShardRouting) {
+            return indexShard.activeInitializingShardsSimpleWeightedIt(
+                weightedRoutingMetadata.getWeightedRouting(),
+                nodes,
+                getWeightedRoutingDefaultWeight(),
+                routingHash
+            );
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/OperationRouting.java:353]: how is this different from `preferAttributesActiveInitializingShardsIt` ?  If we are shuffling and not considering weights for now, we can as well use `preferAttributesActiveInitializingShardsIt`  .

File: server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java (hunk 1 above)
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java:663]: Will remove this. Not needed

File: server/src/main/java/org/opensearch/cluster/routing/OperationRouting.java (hunk 2 above)
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/OperationRouting.java:353]: we are considering weights here with `preferAttributesActiveInitializingShardsIt`

**Review by gbbafna — APPROVED** [review]

**Review by Bukhtawar — APPROVED** [review]

File: server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java [hunk 3]
```diff
@@ -871,6 +887,42 @@ public void testPreferenceSearchWithWeightedRouting() {
...
+        setShardRoutingWeights(weights);
+        String nodeInZoneA = nodeMap.get("a").get(0);
+        String customPreference = randomAlphaOfLength(10);
+
+        assertThrows(
+            PreferenceBasedSearchNotAllowedException.class,
+            () -> internalCluster().client(nodeMap.get("b").get(0))
+                .prepareSearch()
+                .setSize(0)
+                .setPreference(randomFrom("_local", "_only_nodes:" + nodeInZoneA, "_prefer_nodes:" + nodeInZoneA, customPreference))
```
Comment (Bukhtawar) [inline server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java:920]: I don't think we should do randomised testing here. Lets assert each of them. Also why should `customPreference `fails

Comment (andrross) [reply server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java:920]: Agreed. It is generally an anti-pattern [to use randomization for coverage](https://github.com/opensearch-project/OpenSearch/blob/fe8fd67884160e346def8c8503d09fb4b5eb5363/TESTING.md?plain=1#L440).

Comment (anshu1106) [reply server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java]: addressed

**Comment by dblock** [comment]
Needs a rebase.

@Bukhtawar @gbbafna merge this after?

**Pull request merged**